source/dem/read_checkpoint.cc

```diff
@@ -23,13 +25,32 @@ read_checkpoint(
   parallel::distributed::Triangulation<dim>               &triangulation,
   Particles::ParticleHandler<dim>                         &particle_handler,
   std::shared_ptr<Insertion<dim>>                         &insertion_object,
-  std::vector<std::shared_ptr<SerialSolid<dim - 1, dim>>> &solid_surfaces)
+  std::vector<std::shared_ptr<SerialSolid<dim - 1, dim>>> &solid_surfaces,
+  CheckpointControl &checkpoint_controller)
 {
   if (!DEMActionManager::get_action_manager()->check_restart_simulation())
     return;
 
   TimerOutput::Scope timer(computing_timer, "Read checkpoint");
   std::string        prefix = parameters.restart.filename;
+
+  // Load checkpoint controller
+  std::string checkpoint_controller_object_filename =
+    prefix + ".checkpoint_controller";
+  std::ifstream iss_checkpoint_controller_obj(
+    checkpoint_controller_object_filename);
+  boost::archive::text_iarchive ia_checkpoint_controller_obj(
+    iss_checkpoint_controller_obj, boost::archive::no_header);
+  checkpoint_controller.deserialize(ia_checkpoint_controller_obj, 0);
+
+  // Update predix with the checkpoint id
+  prefix =
+    prefix + "_" +
+    Utilities::int_to_string(checkpoint_controller.get_next_checkpoint_id());
+
+  // Need to increment the next checkpoint id for the next checkpoint
+  checkpoint_controller.increment_checkpoint_id();
```

Comment:
Why do we need to increment the checkpoint_id after reading?

Reply:
If you read 0, this means that the last checkpoint was a 0, thus the next one needs to be 1.

Reply:
We just increment after reading so that the next write doesn't erase the last most advance checkpoint in the simulation. 

I'm not sure what you mean... currently the read function is incrementing the checkpoint id.